src/main/core/support/configuration.rs

```diff
@@ -1,9 +1,517 @@
+    #[clap(about = GENERAL_HELP.get("stop_time").unwrap())]
+    stop_time: Option<units::Time<units::TimePrefixUpper>>,
+
+    /// Environment variables passed to all simulated processes (ex: "ENV_A=1;ENV_B=2")
+    #[clap(long, value_name = "env-string")]
+    #[clap(about = GENERAL_HELP.get("environment").unwrap())]
+    environment: Option<String>,
+
+    /// Initialize randomness for each thread using seed N
+    #[clap(long, value_name = "N")]
+    #[clap(about = GENERAL_HELP.get("seed").unwrap())]
+    #[serde(default = "default_some_1")]
+    seed: Option<u32>,
+
+    /// Run concurrently with N worker threads
+    #[clap(long, short = 'w', value_name = "N")]
+    #[clap(about = GENERAL_HELP.get("workers").unwrap())]
+    #[serde(default = "default_some_0")]
+    workers: Option<u32>,
+
+    #[clap(long, value_name = "seconds")]
+    #[clap(about = GENERAL_HELP.get("bootstrap_end_time").unwrap())]
+    #[serde(default = "default_some_time_0")]
+    bootstrap_end_time: Option<units::Time<units::TimePrefixUpper>>,
+
+    /// Log level of output written on stdout. If Shadow was built in release mode, then log
+    /// messages at a level lower than 'info' will always be dropped
+    #[clap(long, short = 'l', value_name = "level")]
+    #[clap(about = GENERAL_HELP.get("log_level").unwrap())]
+    #[serde(default = "default_some_message")]
+    log_level: Option<LogLevel>,
+
+    /// Interval at which to print heartbeat messages
+    #[clap(long, value_name = "seconds")]
+    #[clap(about = GENERAL_HELP.get("heartbeat_interval").unwrap())]
+    #[serde(default = "default_some_time_1")]
+    heartbeat_interval: Option<units::Time<units::TimePrefixUpper>>,
+
+    /// Path to store simulation output
+    #[clap(long, short = 'd', value_name = "path")]
+    #[clap(about = GENERAL_HELP.get("data_directory").unwrap())]
+    #[serde(default = "default_data_directory")]
+    data_directory: Option<String>,
+
+    /// Path to recursively copy during startup and use as the data-directory
+    #[clap(long, short = 'e', value_name = "path")]
+    #[clap(about = GENERAL_HELP.get("template_directory").unwrap())]
+    template_directory: Option<String>,
+}
+
+impl GeneralOptions {
+    /// Replace unset (`None`) values of `base` with values from `default`.
+    pub fn merge(base: &Self, default: &Self) -> Self {
+        let mut base = base.clone();
+        base.merge(default.clone());
+        base
+    }
+}
+
+/// Help messages used by Clap for command line arguments, combining the doc string with
+/// the Serde default.
+static EXP_HELP: Lazy<std::collections::HashMap<String, String>> =
+    Lazy::new(|| generate_help_strs(schema_for!(ExperimentalOptions)));
+
+#[derive(Debug, Clone, Clap, Serialize, Deserialize, Merge, JsonSchema)]
+#[clap(
+    help_heading = "EXPERIMENTAL (Unstable and may change or be removed at any time, regardless of Shadow version)"
+)]
+#[serde(default)]
+pub struct ExperimentalOptions {
```

Comment:
Is there a reason the defaults aren't encoded inline as for GeneralOptions? It's nice to have them near the definitions, and I assume without it the defaults won't show up in `--help`?

Reply:
If the user doesn't provide an "experimental" section, Serde needs to be able to generate a default `ExperimentalOptions` struct, so it needs `impl Default`. The `#[serde(default)]` means that Serde will also be able to use any of those same defaults from the `Default` impl when the user leaves out individual fields (for example `preload_spin_max`).

There's an existing issue for this: https://github.com/serde-rs/serde/issues/1416

> I assume without it the defaults won't show up in --help?

They will still show up in `--help`.